### PR TITLE
[Paper] Regularizing Class-wise Predictions via Self-knowledge Distillation implemented

### PR DESCRIPTION
**Pull request by ashwinvaswani**

### CODE REVIEW

File: KD_Lib/KD/vision/CSKD/sampler.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+from torchvision import transforms, datasets
+from collections import defaultdict
+
+
+class PairBatchSampler(Sampler):
+    def __init__(self, dataset, batch_size, num_iterations=None):
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self.num_iterations = num_iterations
+
```
Comment (Het-Shah) [inline KD_Lib/KD/vision/CSKD/sampler.py:22]: docstrings are missing for many functions.

File: KD_Lib/KD/vision/CSKD/cskd.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    def __init__(self, temp_factor):
+        super(KDLoss, self).__init__()
+        self.temp_factor = temp_factor
+        self.kl_div = nn.KLDivLoss(reduction="sum")
+
+    def forward(self, input, target):
+        log_p = torch.log_softmax(input / self.temp_factor, dim=1)
+        q = torch.softmax(target / self.temp_factor, dim=1)
+        loss = self.kl_div(log_p, q) * (self.temp_factor ** 2) / input.size(0)
+        return loss
```
Comment (Het-Shah) [inline KD_Lib/KD/vision/CSKD/cskd.py:22]: you can prolly shift this to CSKD class itself, we currently use `calculate_kd_loss` for those who have this loss function.

File: KD_Lib/KD/vision/CSKD/cskd.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                train_loss += loss.item()
+
+                with torch.no_grad():
+                    outputs_cls = model(data[batch_size // 2 :])
+                cls_loss = kdloss(outputs, outputs_cls.detach())
+                loss += self.lamda * cls_loss
+                train_cls_loss += cls_loss.item()
+
+                _, pred = torch.max(outputs, 1)
+                # pred = outputs.argmax(dim=1, keepdim=True)
```
Comment (Het-Shah) [inline KD_Lib/KD/vision/CSKD/cskd.py:127]: remove comments like these

**Review by Het-Shah — COMMENTED** [review]
Looks good mostly also add the files for docs otherwise, it will slip out

File: KD_Lib/KD/vision/CSKD/sampler.py (hunk 1 above)
Comment (ashwinvaswani) [reply KD_Lib/KD/vision/CSKD/sampler.py:22]: Will add it

File: KD_Lib/KD/vision/CSKD/cskd.py (hunk 2 above)
Comment (ashwinvaswani) [reply KD_Lib/KD/vision/CSKD/cskd.py:22]: Okay

File: KD_Lib/KD/vision/CSKD/cskd.py (hunk 3 above)
Comment (ashwinvaswani) [reply KD_Lib/KD/vision/CSKD/cskd.py:127]: Sure, will change and update the PR

File: docs/usage/tutorials/CSKD.rst [hunk 4]
```diff
@@ -0,0 +1,33 @@
+======================
+Regularizing Class-wise Predictions via Self-knowledge Distillation using KD_Lib
+======================
+
+To implement the most basic version of knowledge distillation from `Regularizing Class-wise Predictions via Self-knowledge Distillation <https://arxiv.org/abs/2003.13964>`_
+and plot losses
```
Comment (Het-Shah) [inline docs/usage/tutorials/CSKD.rst:6]: Maybe add some explanation, about the paper or just paste the schematic of the whole paper summarizing it

Comment (ashwinvaswani) [reply docs/usage/tutorials/CSKD.rst:6]: Sure. I'll add the summary and key contributions of the paper.

**Comment by ashwinvaswani** [comment]
@Het-Shah Updated with the changes you suggested. Kindly check.

**Review by Het-Shah — APPROVED** [review]
